operator/redisfailover/service/generator.go

```diff
@@ -237,10 +235,10 @@ func generateRedisHeadlessService(rf *redisfailoverv1.RedisFailover, labels map[
 	}
 }
 
-func generateHAProxyService(rf *redisfailoverv1.RedisFailover, labels map[string]string, ownerRefs []metav1.OwnerReference) *corev1.Service {
+func generateHAProxyRedisMasterService(rf *redisfailoverv1.RedisFailover, labels map[string]string, ownerRefs []metav1.OwnerReference) *corev1.Service {
```

Comment:
This service's selector seems wrong to me. It's picking up the slave haproxy pod as well:

```
$ kubectl --context kind-test --namespace redis-operator-test get service/rfrm-haproxy-cache -o json | jq .spec.selector
{
  "app.kubernetes.io/component": "redis",
  "app.kubernetes.io/managed-by": "redis-operator",
  "redisfailovers.databases.spotahome.com/component": "haproxy",
  "redisfailovers.databases.spotahome.com/name": "cache"
}

$ kubectl --context kind-test --namespace redis-operator-test get pods --selector='app.kubernetes.io/component=redis,app.kubernetes.io/managed-by=redis-operator,redisfailovers.databases.spotahome.com/component=haproxy,redisfailovers.databases.spotahome.com/name=cache'
NAME                                  READY   STATUS    RESTARTS   AGE
rfrm-haproxy-cache-756c768f75-tcfzq   1/1     Running   0          7m8s
rfrs-haproxy-cache-59b8bd575-n84g9    1/1     Running   0          7m8s
```

...which causes writes through the master service to frequently fail about half the time.

```
root@utility-756f449d6c-5mzxt:/# while true; do redis-cli -h rfrm-haproxy-cache incr counter; sleep 1; done
(integer) 1
(integer) 2
(error) READONLY You can't write against a read only replica.
(integer) 3
(error) READONLY You can't write against a read only replica.
(integer) 4
(integer) 5
(integer) 6
(error) READONLY You can't write against a read only replica.
```

Reply:
yes, that is correct. label added

Reply:
> My thoughts is that we need to change app.kubernetes.io/component for HAProxy someday. Probably new issue should be opened

FWIW, I see 0c7d96b42a88e5bfa5d924472123e8b41bd26120 fixes the HAProxy RedisMasterDeployment's  selectors as well. [Deployment selectors are also immutable](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#selector). Rolling this out will require human intervnetion - remove the old deployment and let the operator generate a new one with the correct labels. Should we take this opportunity to fix ALL of the HAProxy labels and selectors in one pass?